src/pczt/verify.rs

```diff
@@ -0,0 +1,190 @@
+use crate::{keys::FullViewingKey, value::ValueCommitment, Note, ViewingKey};
+
+impl super::Spend {
+    /// Verifies that the `cv` field is consistent with the note fields.
+    ///
+    /// Requires that the following optional fields are set:
+    /// - `value`
+    /// - `rcv`
+    pub fn verify_cv(&self) -> Result<(), VerifyError> {
+        let value = self.value.ok_or(VerifyError::MissingValue)?;
+        let rcv = self
+            .rcv
+            .clone()
+            .ok_or(VerifyError::MissingValueCommitTrapdoor)?;
+
+        let cv_net = ValueCommitment::derive(value, rcv);
+        if cv_net.to_bytes() == self.cv.to_bytes() {
+            Ok(())
+        } else {
+            Err(VerifyError::InvalidValueCommitment)
+        }
+    }
+
+    /// Returns the [`ViewingKey`] to use when validating this note.
+    ///
+    /// Handles dummy notes when the `value` field is set.
+    fn vk_for_validation(
+        &self,
+        expected_fvk: Option<&FullViewingKey>,
+    ) -> Result<ViewingKey, VerifyError> {
+        let vk = self
+            .proof_generation_key
+            .as_ref()
+            .map(|proof_generation_key| proof_generation_key.to_viewing_key());
+
+        match (expected_fvk, vk, self.value.as_ref()) {
+            // Dummy notes use random keys, which must be provided.
+            (_, Some(vk), Some(value)) if value.inner() == 0 => Ok(vk),
+            (_, None, Some(value)) if value.inner() == 0 => {
+                Err(VerifyError::MissingProofGenerationKey)
+            }
+            // If the `proof_generation_key` field has been pruned, assume the caller
+            // provided the correct FVK.
+            (Some(expected_fvk), None, _) => Ok(expected_fvk.vk.clone()),
+            // This is not a dummy note; if the FVK field is present, it must match.
+            (Some(expected_fvk), Some(vk), _)
+                if vk.ak == expected_fvk.vk.ak && vk.nk == expected_fvk.vk.nk =>
+            {
+                Ok(vk)
+            }
+            (Some(_), Some(_), _) => Err(VerifyError::MismatchedFullViewingKey),
+            (None, Some(vk), _) => Ok(vk),
+            (None, None, _) => Err(VerifyError::MissingProofGenerationKey),
+        }
```

Comment:
The following is more obviously just logical unification with the `Some(_), Some(fvk), Some(0)` case treated as an exception (which is what the docs for `verify_nullifier` and `verify_rk` suggest to me that it should be):
```suggestion
        match (expected_fvk, vk, self.value.as_ref()) {
            (Some(expected_fvk), Some(vk), _)
                if vk.ak == expected_fvk.vk.ak && vk.nk == expected_fvk.vk.nk =>
            {
                Ok(vk)
            }
            // `expected_fvk` is ignored if the spent note is a dummy note.
            (Some(_), Some(vk), Some(value)) if value.inner() == 0 => Ok(vk),
            (Some(_), Some(_), _) => Err(VerifyError::MismatchedFullViewingKey),
            (Some(expected_fvk), None, _) => Ok(expected_fvk),
            (None, Some(vk), _) => Ok(vk),
            (None, None, _) => Err(VerifyError::MissingProofGenerationKey),
        }
```